temporian/core/data/schema.py

```diff
@@ -0,0 +1,134 @@
+# Copyright 2021 Google LLC.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Schema and related classes."""
+
+from __future__ import annotations
+
+from typing import List, Tuple, TYPE_CHECKING, Dict, Union
+
+from dataclasses import dataclass
+from temporian.core.data.dtype import DType, IndexDType
+
+if TYPE_CHECKING:
+    from temporian.core.operators.base import Operator
+
+
+@dataclass
+class FeatureSchema:
+    name: str
+    dtype: DType
+
+    def __repr__(self) -> str:
+        return f"({self.name!r}, {self.dtype})"
+
+
+@dataclass
+class IndexSchema:
+    name: str
+    dtype: IndexDType
+
+    def __repr__(self) -> str:
+        return f"({self.name!r}, {self.dtype})"
+
+
+class Schema:
+    """A schema defines the type of data in an event set.
+
+    A schema does not contain any actual data.
+    A schema can be shared by multiple nodes.
+
+    Attributes:
+        features: List of feature names and types.
+        indexes: List of index names ans types.
+        is_unix_timestamp: Whether values correspond to Unix timestamps.
+    """
+
+    def __init__(
+        self,
+        features: Union[List[FeatureSchema], List[Tuple[str, DType]]],
+        indexes: Union[List[IndexSchema], List[Tuple[str, IndexDType]]],
+        is_unix_timestamp: bool,
+    ):
+        def normalize_feature(x):
+            if isinstance(x, FeatureSchema):
+                return x
+            assert len(x) == 2
+            return FeatureSchema(x[0], x[1])
+
+        def normalize_index(x):
+            if isinstance(x, IndexSchema):
+                return x
+            assert len(x) == 2
+            return IndexSchema(x[0], x[1])
```

Comment:
could these be defined as private methods in the class instead of each time the constructor runs?

Reply:
I would not turn them into methods as they don't use self.
I've moved them at the end of the file.